model-optimizer/mo/back/preprocessing.py

```diff
@@ -0,0 +1,383 @@
+                    suitable_params.append(param_info)
+            elif layout_item.get('source_layout'):
+                if check_suitable_for_reverse(Layout(layout_item['source_layout']), ov_input):
+                    suitable_params.append(param_info)
+            continue
+
+        try:
+            layout_values = find_channels_dimension(shape=ov_input.get_partial_shape(),
+                                                    num_channels=3,
+                                                    name=first_name,
+                                                    layout_values=layout_values)
+        except Error as e:
+            log.debug('Reverse input channels guess did not succeed {}'.format(e))
+        else:
+            layout = layout_values[first_name].get('source_layout')
+            if layout and check_suitable_for_reverse(Layout(layout), ov_input):
+                suitable_params.append(param_info)
+
+    if len(suitable_params) < len(all_params):
+        log.error('Network has {} inputs overall, but only {} of them are suitable for input channels reversing.\n'
+                  'Suitable for input channel reversing inputs are 4-dimensional with 3 channels\nAll inputs: {}\n'
+                  'Suitable inputs {}'.format(len(all_params), len(suitable_params), all_params, suitable_params),
+                  extra={'is_warning': True})
+    return suitable_params
+
+
+def apply_preprocessing(ov_function: Function, argv: argparse.Namespace):
+    """
+    Applies pre-processing of model inputs by adding appropriate operations
+    On return, 'ov_function' object will be updated
+    :param: ov_function OV function for applying mean/scale pre-processing
+    :param: argv Parsed command line arguments
+    """
+    prep = PrePostProcessor(ov_function)
+
+    if 'mean_scale_values' in argv and argv.mean_scale_values:
+        mean_scale_values = argv.mean_scale_values
+    else:
+        mean_scale_values = {}
+
+    mean_scale_values = update_mean_scale_to_dict(input_nodes=ov_function.inputs,
+                                                  mean_scale_val=mean_scale_values,
+                                                  scale=argv.scale)
+
+    layout_values = {}
+    if 'layout_values' in argv and argv.layout_values:
+        layout_values = argv.layout_values
+
+    if '' in layout_values:
+        if len(ov_function.inputs) > 1:
+            input_names = [list(ov_input.get_tensor().get_names())[0] for ov_input in ov_function.inputs]
+            raise Error('Layout without name can be specified for models with only one input, '
+                        'but provided model has {} inputs: \'{}\'. '
+                        'Please specify explicitly input/output name for --layout option'
+                        .format(len(input_names), input_names))
+        layout_values = {
+            list(ov_function.input().get_tensor().get_names())[0]: {
```

Comment:
can we have a case when we have many names for the single input and layout dictionary uses a name different from mean_scale_val dictionary? Are we sure that there will be no name mismatch?

Reply:
Yes, this is covered by `check_keys_valid(keys=layout_values.keys()` (few lines below)